### ISSUE TITLE
Possible regressions with 0.25

### ISSUE THREAD

**Issue by willingc** [nteract/archived-desktop-app#5283]
@captainsafia I'm seeing some issues with:
- closing/quitting the desktop app (had to `ps -ax | grep nteract` and then kill)
- switching of editors resulting in blank page
- individual windows will not close

I was checking if the Export -> PDF worked and ran into this.

**Comment by codepatel** [comment]
shall I wait for evaluating v0.25 until these issues are resolved?

**Comment by willingc** [comment]
@codepatel I was seeing these issues on macOS from running from the dmg. I will be trying to reproduce this today/tomorrow. If you want to give it a try, I would appreciate it and we would welcome more feedback in this issue. Thanks.

**Comment by kubark42** [comment]
@willingc I do not see anything which strikes me as unusual on a Mac Air with macOS 10.15.6. I am running v0.25 which installed automatically as an upgrade from 0.24.1 (i.e. so not running from a DMG).

However, I am not sure that I have fully understood the 2nd issue, so I cannot guarantee that I have tested it correctly.

**Comment by jarmokivekas** [comment]
>
>     * switching of editors resulting in blank page
> 

^^ I'm seeing this on v0.25.0 ~and current master.~ (actually just the release, master is fine)

1. open any notebook
1. Preference > Editor Type > Monaco 
1. entire page goes white

I'm running the .deb install package on Ubuntu.

**Comment by vivek1729** [comment]
@willingc , @captainsafia  I am seeing the issue as well on the Windows App for the latest release. However, when I try to run the desktop app locally from the `main` branch, I don't see any issues. Can you help me with the instructions to do a production build for the desktop app locally? 
I feel some difference in the way the app is packaged in prod vs dev might have triggered this issue.

**Comment by vivek1729** [comment]
Pre-requisite: I downloaded the nteract setup for the desktop release version 0.25.0 from nteract.io
Took a closer look at the issue and I think I was able to root-cause the issue when switching editors. Looking at the console I see the following error:

![image](https://user-images.githubusercontent.com/1658576/92779509-f74d5980-f356-11ea-9c86-2655c9db1075.png)

The [call ](https://github.com/nteract/nteract/blob/main/packages/monaco-editor/src/MonacoEditor.tsx#L129) happens within the `calculateHeight` method which dynamically resizes the editor container when the component is mounted. Since the call errors out, it throws an uncaught exception which ends up rendering a blank page.

I am surprised to see the error at this call `this.editor.getConfiguration().lineHeight`. I checked the [API docs](https://github.com/microsoft/monaco-editor/blob/dc793af0020b9878b798b45fdcaf31e1b263c40b/website/playground/monaco.d.ts.txt) of the [0.18.1](https://github.com/microsoft/monaco-editor/releases/tag/v0.18.1). release for the `monaco-editor` and I can confirm that the [`getConfiguration`](https://github.com/microsoft/monaco-editor/blob/dc793af0020b9878b798b45fdcaf31e1b263c40b/website/playground/monaco.d.ts.txt#L4007) method is defined for the Editor contract. On further investigation, I tried to spy on the Editor APIs for this desktop release and it seems like it is picking up the API definitions from a later version of the `monaco-editor`. I could confirm that because `getConfiguration` API was removed in the newer version of Monaco Editor. I also saw newer API methods (not defined in 0.18.1 version) like [`getRawOptions`](https://microsoft.github.io/monaco-editor/api/interfaces/monaco.editor.istandalonecodeeditor.html#getrawoptions) when I looked at the editor APis in the console.


On removing the `getConfiguration` call in the sources of the desktop App, I am able to load the editors and switch between them without hiccups. The app also doesn't freeze and closes normally when we try to quit.

**TLDR:**
Looks like the latest version of desktop app is picking up a newer version of `monaco-editor` module when it is packaged for production and it is not respecting the version that we've specified in `package.json`  i.e. 0.18.1. 

It is interesting because this issue doesn't repro when I try to run the app locally from the `main` branch using the `dev` environment.  

Referencing the [PR ](https://github.com/nteract/nteract/pull/5279) automatically created to automatically upgrade `monaco-editor` node module.

cc @willingc @captainsafia

**Comment by willingc** [comment]
Ping @captainsafia who has done production releases. Sorry I have not done them yet.

**Comment by captainsafia** [comment]
The RELEASING.md documentation in the root of the repo contains docs on this. Should just need to run `yarn dist` then find the executable in `application/desktop/dist`.

**Comment by vivek1729** [comment]
Thanks @captainsafia. I consistently get a heap allocation error when I try to do `yarn dist`. I am on a windows 10 machine. I was able to unblock myself by setting the `NODE_OPTIONS` environment variable. On Windows, I did `set NODE_OPTIONS=--max_old_space_size=8172`. Had a follow-up question just to confirm that I validate thoroughly. The `dist` folder has the following structure:
![image](https://user-images.githubusercontent.com/1658576/93649863-d07bcc80-f9c1-11ea-9f50-a3edf15ea51f.png)

The `win-unpacked` folder has the following contents:
![image](https://user-images.githubusercontent.com/1658576/93649917-f5703f80-f9c1-11ea-8465-6c38e715ab1c.png)

Do you think it is sufficient for me to test the `nteract.exe` within the `win-unpacked` folder in this scenario where I want to validate the prod release candidate?

**Comment by vivek1729** [comment]
To verify the electron app, I ran `yarn dist` on two commit points:
- I checked out the [commit](https://github.com/nteract/nteract/commit/b87e929837496ed6f45e8b4ec48add699709fd7d) pertaining to the August release
- I checked against the latest pull from `main`

For both these commits, I was able to successfully run `nteract.exe` in the `win-unpacked` folder and also run the app without hiccups after installing it from the outputed `nteract Setup 0.25.0.exe`. I could also confirm that the setups work on another machine where I did not build the project itself to generate these output executables. 

TLDR; I was able to switch editors and use monaco, get completions on the latest main and the commit from the August release. I am not sure what might be the root cause at this point.

It will be great if someone else confirms this as well so that we can ensure we have ruled out any inconsistencies that might arise from a different machine setup. Although, I think that the output executables generated should be agnostic of the build machine configuration entirely.

@captainsafia, Can you verify if the output from `yarn dist` (from latest `main`) works on your machine as well and that you are able to switch editors without any issues?

**Comment by captainsafia** [comment]
OOOOOOOOOOK! Thanks folks for all your patience on this thread. Been a little busy with day job/writing/variety of other shenanigans.

@vivek1729 You followed the correct steps for producing a local production build. Having to increase the allocated heap size is not-problematic since the Electron build does consume quite a bit of memory.

I was able to reproduce the issue on the v0.25 release on macOS. I was not able to reproduce this issue on Windows. 

> closing/quitting the desktop app (had to ps -ax | grep nteract and then kill)
> switching of editors resulting in blank page
> individual windows will not close

I wasn't able to reproduce these issues on Windows, but I was able to reproduce on macOS. It appears that the unhandled exception from the monaco-editor completely borks all the event handlers in the web page. Users are unable to close the app page in this case because we register event handlers on close inside the JS on the page (as opposed to Electron).

So, in summary, the root cause of the issues above appears to be unhandled exception with the undefined `getConfiguration` API.

Unfortunately for us, the issue appears to repro inconsistently. I was able to repro on macOS in v0.25 that was auto-updated. @willingc was able to reproduce in a fresh app on the DMG. And it appears that it repros on Linux as well.

As to why this happens, it appears that the `getConfiguration` API was deprecated in 0.19 ([ref](https://github.com/microsoft/monaco-editor/blob/398fd0eee67939d2dd6d0efe211ee79113ac8da7/CHANGELOG.md#breaking-changes-2)). I was a bit suspicious about this so I cracked upon the build application and inspected the `app.js` file bundled in the app.

```
$ cd /Applications/nteract.app/Contents/Resources
$ asar extract app.asar nteract-built
$ code nteract-built
```

Sure enough, a lookup for `getConfiguration` in that file yields nothing. But `getRawOptions` is defined.

So somehow, the incorrect version of monaco-editor is being bundled into some applications. Why this is happening? I'm not sure. The inconsistency in reproducibility we are seeing makes this difficult to track down. Next thing I'll try is to see if the issue reproduces on a fresh build (yarn dist) on macOS.

@vivek1729 Any intuition on why some builds bundle the incorrect version of monaco-editor?

**Comment by vivek1729** [comment]
@captainsafia , thanks a lot for taking a look. That was exactly my [observation](https://github.com/nteract/nteract/issues/5283#issuecomment-690605324) when I tried to root-cause it in the last few days.

It's fascinating that the package is auto-updated for Mac and not on Windows. It's also possible that this happens for some other packages too but we have not yet encountered run time errors with them. I double checked the [`package.json`](https://github.com/nteract/nteract/blob/b87e929837496ed6f45e8b4ec48add699709fd7d/packages/monaco-editor/package.json#L28) and the [`yarn.lock`](https://github.com/nteract/nteract/blob/main/yarn.lock#L12047) file for monaco-editor and sure enough the version is pinned so it should not automatically get a newer version.

I don't have a hypothesis at this point but I think it might be worthwhile to understand at which stages in the release process are node modules installed/updated. Is it possible that there is a difference in the way some webpack/electron configuration is interpreted on different OS? For eg:  Could it be the case that some environment variable is not getting set correctly for all the OS targets?

Doing a clean `yarn dist` might be a good idea. I am not sure if that command installs the node modules again. So, if we can confirm that we had the right version of `monaco-editor` package installed when we did `yarn install` or `yarn start`, then `yarn dist` should just pick up the same node modules right?

**Comment by captainsafia** [comment]
> It's fascinating that the package is auto-updated for Mac and not on Windows. It's also possible that this happens for some other packages too but we have not yet encountered run time errors with them.

Auto-update is supported on all operating systems, which is what makes this more puzzling. In addition to the fact that @willingc reproduced this on a fresh `.app` file from a DMG on macOS.

My thinking about how to approach this at this point:
1. Generate the distributed builds for all operating systems (have to do this on macOS because you can only make macOS builds on macOS devices)
2. Inspect the bundled app.js file in each and confirm that the appropriate version of monaco-editor is shipped
3. If so, great, something might've gonna awry with the last release
4. If not, investigate why the particular incorrect builds are created

If the October release contains the same issue even after validating that the right bits are included as part of #3, then it very well maybe something about the target machine and not the bundle.

> Doing a clean yarn dist might be a good idea. I am not sure if that command installs the node modules again. So, if we can confirm that we had the right version of monaco-editor package installed when we did yarn install or yarn start, then yarn dist should just pick up the same node modules right?

Yep. For what you are proposing (same as step #1 above), we'll need to do `yarn clean && yarn && yarn build:all:ci && yarn dist`.

**Comment by vivek1729** [comment]
@willingc , I can confirm that the editor switching works correctly with the `0.25.1` release :)

**Comment by captainsafia** [comment]
Yep! Meant to respond back to this thread after shipping the release this week.

I'll close this issue for now. Export to PDF still needs work but we have another issue tracking that.
